aleph-client/src/contract/mod.rs

```diff
@@ -0,0 +1,144 @@
+/// Utilities for writing contract wrappers.
+///
+/// The functions in this module simplify parsing the data types returned by contract calls.
+pub mod util;
+
+use std::{
+    fmt::{Debug, Formatter},
+    fs::File,
+};
+
+use ac_primitives::ExtrinsicParams;
+use anyhow::{anyhow, Context, Result};
+use contract_metadata::ContractMetadata;
+use contract_transcode::{ContractMessageTranscoder, Value};
+use ink_metadata::{InkProject, MetadataVersioned};
+use serde_json::{from_reader, from_str, from_value, json};
+use sp_core::{crypto::AccountId32, Pair};
+use substrate_api_client::{compose_extrinsic, GenericAddress, XtStatus};
+
+use crate::{try_send_xt, AnyConnection, Connection, SignedConnection};
+
+/// Represents a contract instantiated on the chain.
+pub struct ContractInstance {
+    address: AccountId32,
+    ink_project: InkProject,
+}
+
+impl ContractInstance {
+    const MAX_READ_GAS: u64 = 500000000000u64;
+    const MAX_GAS: u64 = 10000000000u64;
+
+    /// Creates a new contract instance under `address` with metadata read from `metadata_path`.
+    pub fn new(address: AccountId32, metadata_path: &str) -> Result<Self> {
+        Ok(Self {
+            address,
+            ink_project: load_metadata(metadata_path)?,
```

Comment:
My scout rules are protesting against reading a file in a constructor, and instead I would expect `new` to have a signature `pub fn new(address: AccountId32, metadata_path: &InkProject) -> Self` (or possibly just `InkProject`).

Reply:
I think it's perfectly file to do operations that can fail in the constructors. I think this rule comes from the old C-days when there was not any exceptional safety of code so it looked like `FileHandle x = someClassConstructor(file); x.load(); if x.okay() {...} x.close();`  etc  Now we have `Result`s so it's fine for a constructor to fail

Reply:
yeah, but you pass the whole responsibility of checking for file and reading it to a constructor of a contract representation, which seems to me a bit bad

Reply:
"where to get such a thing" - under the path that they are passing to this constructor xD

nevertheless, I won't be protesting longer, go ahead, but know that I'm not so convinced :P

Reply:
Under that path there is a file, not an `InkProject` :P

I don't think it matters, and we can refactor later if we need it.